crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -562,33 +564,104 @@ impl SnapshotRepository {
-            let region = aws_config::load_defaults(BehaviorVersion::v2024_03_28())
-                .await
-                .region()
-                .context("Unable to determine AWS region")?
-                .to_string();
-            debug!(?profile, ?region, "Using AWS SDK credentials provider");
-            AmazonS3Builder::new()
-                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
-                    credentials_provider: DefaultCredentialsChain::builder().build().await,
-                }))
-                .with_region(region)
+        let builder = match &options.aws_profile {
+            Some(profile) => {
+                debug!(profile, "Using AWS profile for snapshot repository access");
+                let sdk_config = aws_config::defaults(BehaviorVersion::v2024_03_28())
+                    .profile_name(profile)
+                    .load()
+                    .await;
+
+                let region = DefaultRegionChain::builder()
+                    .profile_name(profile)
+                    .build()
+                    .region()
+                    .await
+                    .context("Unable to determine region from profile")?;
+
+                debug!(?region, ?profile, "Using AWS SDK credentials provider");
+                let creds = DefaultCredentialsChain::builder()
+                    .profile_name(profile)
+                    .region(region.clone())
+                    .build()
+                    .await;
+
+                let builder = AmazonS3Builder::new()
+                    .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                        credentials_provider: creds,
+                    }))
+                    .with_region(region.to_string());
+
+                if let Some(endpoint_url) = sdk_config.endpoint_url() {
+                    debug!(endpoint_url, "Using custom AWS endpoint from profile");
+                    builder.with_endpoint(endpoint_url)
+                } else {
+                    builder
+                }
+            }
+            None => {
+                let sdk_config = aws_config::load_defaults(BehaviorVersion::v2024_03_28()).await;
+
+                let region = sdk_config
+                    .region()
+                    .context("Unable to determine AWS region")?
+                    .to_string();
+
+                let builder = AmazonS3Builder::new().with_region(region.clone());
+                if options.aws_access_key_id.is_none() {
+                    debug!(?region, "Using AWS SDK credentials provider");
+                    builder.with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                        credentials_provider: DefaultCredentialsChain::builder().build().await,
+                    }))
+                } else {
+                    builder
+                }
+            }
+        };
+
+        let builder = if let Some(region) = &options.aws_region {
```

Comment:
These would be a great candidate for a macro, I think, but I don't want to start this right now 😅